sycl/plugins/esimd_emulator/pi_esimd_emulator.hpp

```diff
@@ -104,48 +110,96 @@ struct _pi_mem : _pi_object {
 
   pi_context Context;
 
+  // To be used for piEnqueueMemBufferMap
   char *MapHostPtr = nullptr;
 
-  // Mutex for load/store accessing
+  // Mutex for load/store accessing to CM-managed surfaces from memory
+  // instrinsic implementations
   std::mutex mutexLock;
 
-  // Surface index used by CM
-  int SurfaceIndex;
+  // Surface index
+
+  unsigned int SurfaceIndex;
+  // Supplementary data to keep track of the mappings of this memory
+  // created with piEnqueueMemBufferMap
+  struct Mapping {
+    // The offset in the buffer giving the start of the mapped region.
+    size_t Offset;
+    // The size of the mapped region.
+    size_t Size;
+  };
+
+  // The key is the host pointer representing an active mapping.
+  // The value is the information needed to maintain/undo the mapping.
+  // TODO : std::unordered_map is imported from L0.
+  // Use std::stack for strict LIFO behavior checking?
+  std::unordered_map<void *, Mapping> Mappings;
+  // Supporing multi-threaded mapping/unmapping calls
+  std::mutex MappingsMutex;
 
   virtual ~_pi_mem() = default;
 
   _pi_mem_type getMemType() const { return MemType; };
 
 protected:
   _pi_mem(pi_context ctxt, char *HostPtr, _pi_mem_type MemTypeArg,
-          int SurfaceIdxArg)
+          unsigned int SurfaceIdxArg)
       : Context{ctxt}, MapHostPtr{HostPtr},
         SurfaceIndex{SurfaceIdxArg}, MemType{MemTypeArg} {}
 
 private:
   _pi_mem_type MemType;
 };
 
+// TODO: Merge cm_buffer_ptr_slot and cm_image_ptr_slot into one
+// struct
+struct cm_buffer_ptr_slot {
+  // 'UP' means 'User-Provided' in CM Lib - corresponding to
+  // Host-created buffer in SYCL
+
+  enum type { type_none, type_regular, type_user_provided };
+  type tag = type_none;
+
+  union {
+    cm_support::CmBuffer *RegularBufPtr = nullptr;
+    cm_support::CmBufferUP *UPBufPtr;
+  };
+};
+
+struct cm_image_ptr_slot {
+  // 'UP' means 'User-Provided' in CM Lib - corresponding to
+  // Host-created image in SYCL
+
+  enum type { type_none, type_regular, type_user_provided };
```

Comment:
Copy paste, but common parts could be inherited.

Reply:
Like commented in Line 154, these structures will be merged into one.